server/src/main/java/io/druid/server/BrokerQueryResource.java

```diff
@@ -102,4 +120,43 @@ public Response getQueryTargets(
       return context.gotError(e);
     }
   }
+
+  @Override
+  protected Query prepareQuery(Query query)
+  {
+    return rewriteDataSources(query);
+  }
+
+  private Query rewriteDataSources(Query query)
+  {
+    DataSource dataSource = query.getDataSource();
+    if (dataSource instanceof RegexDataSource) {
+      List<Matcher> matchers = Lists.newArrayList();
+      for (String pattern : dataSource.getNames()) {
+        matchers.add(Pattern.compile(pattern).matcher(""));
+      }
+      Set<String> found = Sets.newLinkedHashSet();
+      for (DruidServer server : serverInventoryView.getInventory()) {
+        for (DruidDataSource source : server.getDataSources()) {
+          if (!found.contains(source.getName())) {
+            for (Matcher matcher : matchers) {
+              if (matcher.reset(source.getName()).matches()){
+                found.add(source.getName());
+                break;
+              }
+            }
+          }
+        }
+      }
+      if (found.size() == 1) {
+        query = query.withDataSource(new TableDataSource(Iterables.getOnlyElement(found)));
+      } else {
+        query = query.withDataSource(new UnionDataSource(TableDataSource.of(found)));
+      }
+    } else if (dataSource instanceof QueryDataSource) {
+      Query subQuery = rewriteDataSources(((QueryDataSource) query).getQuery());
```

Comment:
`(QueryDataSource) query` is a bad cast, this line should probably be:

```java
Query subQuery = rewriteDataSources(((QueryDataSource) dataSource).getQuery());
```

Please also add tests for this case.